Navigation block: fix padding on mobile overlay when global padding is 0

MaggieCabrera opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

This PR expands on the fix made by https://github.com/WordPress/gutenberg/pull/43576 to cover the cases when themes don't declare a global top padding for their site (or they declare it as 0px). This makes sure that the padding value is not 0.

Fixes https://github.com/WordPress/gutenberg/issues/52233

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

The original fix was adding a 2rem fallback if the global padding is not set by the theme. That fallback is probably only working for classic themes, since when block themes don't declare a global padding, the value of the CSS variable resolves to 0px, causing this issue on the overlay menu:

![Screen Capture on 2023-08-16 at 11-21-11](https://github.com/WordPress/gutenberg/assets/3593343/4850abd4-e9da-4057-ba18-501f7f78d057)

There is a use case for themes with no global padding-top when the theme wants to have a different color on the header than the body, where the theme uses group padding instead to separate the menu from the top of the viewport.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Using `max()`, we make sure that the highest value of the pair is always selected, so if the global padding is 0, it will default to 2rem instead.

There are two caveats to this implementation:

1. If the theme defines a non 0 global padding that is lower than 2rem, the 2rem value will prevail. This causes a slight visual jump when opening the overlay
2. This solution doesn't work if theme defines the global padding as a unitless 0. `max(`) needs the values to have units, else they won't work.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

In your theme.json file, define the global padding as either 

```
{
    "styles": {
               "spacing": {
			"padding": {
				"right": "var(--wp--preset--spacing--60)",
				"left": "var(--wp--preset--spacing--60)"
			}
		},
    }
}
```

or

```
{
    "styles": {
               "spacing": {
			"padding": {
                                "top": "0px",
				"right": "var(--wp--preset--spacing--60)",
				"left": "var(--wp--preset--spacing--60)",
                                "bottom": "0px",
			}
		},
    }
}
```

Check the frontend both on a logged in and logged out state, and notice that the burger icon is separated 2rem from the top of the viewport.

Also, check that if the value for the global padding is bigger than 2rem, that's the value that's been applied for the overlay too

MaggieCabrera committed:
force min value for padding to be 2rem

MaggieCabrera commented:
@jasmussen I'm trying this, and it doesn't work after building the CSS, but if I write the same CSS manually in the browser inspector, it works as intended!!! I don't understand what's going on here.

Also, we need to consider that this solution will make any themes that have a padding set that is more than 0 but less than 2rem to jump between closed/open. It's not ideal but I think it's better than what we have and we have the expectation of a better solution to come with https://github.com/WordPress/gutenberg/issues/43852

MaggieCabrera commented:
> @jasmussen I'm trying this, and it doesn't work after building the CSS, but if I write the same CSS manually in the browser inspector, it works as intended!!! I don't understand what's going on here.
> 

It fails if the padding is declared as 0 without any unit, then the max() doesn't know what to do. And the compiler is merging the 4 padding values into the shorthand so having one wrong value breaks the whole thing :(

jasmussen commented:
> And the compiler is merging the 4 padding values into the shorthand so having one wrong value breaks the whole thing

If that's the SCSS thing, then there should be a way around that. For example if I want a variable explicitly output so that `calc` can work with it, then it would be something like `width: calc( 100% - #{ $grid-unit-10 } );`

Let me take a look.

MaggieCabrera commented:
Is it safe to assume that theme authors should add units for this to work? I can make the fix for Blue note, I'm not sure how many other themes would make the same mistake of adding the unitless 0

jasmussen commented:
I think I may have a fix in https://codepen.io/joen/pen/RwEbgQv?editors=1100

Specifically, it _seems_ like `calc( 0 * 1px)` returns `0px`.

MaggieCabrera commented:
> I think I may have a fix in https://codepen.io/joen/pen/RwEbgQv?editors=1100
> 
> Specifically, it _seems_ like `calc( 0 * 1px)` returns `0px`.

It breaks for --global-padding: 4rem;

jasmussen commented:
Oh right. 🤔 

Maybe there is no obvious fix other than addressing that theme.json styleable overlay.

MaggieCabrera commented:
> Oh right. 🤔
> 
> Maybe there is no obvious fix other than addressing that theme.json styleable overlay.

I think this here is the best we can do for now!

MaggieCabrera commented:
I'm gonna go for a bit but I'll be back and write a proper description and test instructions

MaggieCabrera committed:
fallback for when the css variables are not defined

scruffian commented:
I added a commit to use clamp. This means that we can set a min and max, but also allow the size to be the correct one if that's an option. Feel free to remove it if you don't like that option.

WRT the `0` values, I think we should probably consider adding `px` to them in the theme.json output...

MaggieCabrera commented:
https://www.oddbird.net/2022/08/04/zero-units/ I’m not sure adding the unit is a good idea

scruffian commented:
In that case we should probably try to educate peopler about why they should use `0px`!

scruffian committed:
Allow the padding to be smaller than 2rem

tellthemachines reviewed (approved):
Code looks good and the change is working well for me!

One thing I noticed in testing (unrelated to this PR) is that a 46px margin-top is added to `wp-block-navigation__responsive-dialog` on the front end when the admin bar is detected, but the nav overlay hides the admin bar. Might be that z-indexes were messed with at some point? I'm guessing the intention was for the admin bar to show when the overlay is open?

tellthemachines commented:
The e2e failure looks legit and I can reproduce it locally on this branch, but the behaviour tested hasn't changed from trunk to this branch. I suspect something in the spacing change causes hover on the page link button to be activated, and the hover color is being read as the actual link color 🤔 

Edit: failure artifact confirms the problem is hover style:

![test-failed-1](https://github.com/WordPress/gutenberg/assets/8096000/740faa2a-6f54-451f-ad1c-2f7a697aaefd)

getdave commented:
I tested this locally using the following command 

```
npm run test:e2e:playwright -- -g "Navigation text does not inherit the link color from the theme/group" --debug
```

If you step through the test slowly using the debugger it passes correctly. However running at full speed (without debug flag) causes it to fail.

Reverting the changeset in this PR and then re-running the test will cause it to **pass**:

```
git checkout trunk -- packages/block-library/src/navigation/style.scss
```

getdave committed:
Add fix to avoid trigger hover state on links when overlay opens

getdave commented:
Ok I believe I've fixed [this](https://github.com/WordPress/gutenberg/pull/53725#issuecomment-1732933808). The issue was that when the mobile overlay menu was toggled open the mouse pointer was sitting directly on top of one of the links in the overlay. This caused the `:hover` to trigger and thus when Playwright queried the element's CSS it was found to have an unexpected color.

I've fixed this by shifting the mouse down and to the right by a large value in order to ensure that the mouse cannot accidentally trigger any state changes.

MaggieCabrera commented:
> Ok I believe I've fixed this. The issue was that when the mobile overlay menu was toggled open the mouse pointer was sitting directly on top of one of the links in the overlay. This caused the `:hover` to trigger and thus when Playwright queried the element's CSS it was found to have an unexpected color.
> 
> I've fixed this by shifting the mouse down and to the right by a large value in order to ensure that the mouse cannot accidentally trigger any state changes.

Wow, nice detective skills there! Thank you for picking this one up

getdave commented:
Auto merging once tests pass 👍

MaggieCabrera commented:
I added the backport label since this is affecting TT2 and TT4 @mikachan

getdave commented:
Given that there are _no_ PHP changes in this PR these constantly failing PHP tests seem to be anomalous. Agreed?

Re-running again in the hope that they pass this time. There is no error, just that the test fail to run 🤦

mikachan commented:
I just cherry-picked this PR to the release/16.7 branch to get it included in the next release: 7648f9b111
